sql/schemachanger: support for create sequence inside the declarative schema changer

fqazi opened the pull request:
This patch implements CREATE SEQUENCE in the declarative schema changer. The first three commits can be ignored and are included in (https://github.com/cockroachdb/cockroach/pull/104348, which should be reviewed and merged first). This patch will:

- Skip validation/backfill for descriptors in an added state
- Fixes prefix resolution creating new objects, where two-part names were not handled properly before
- Adds support for CREATE sequence in the declarative schema changer

Fixes: #104351

fqazi committed:
sql/schemachanger: fix prefix resolution in declarative schema changer

Previously, when trying to resolve prefixes for newly
created objects, we didn't handle two parts names
properly in the declarative schema changer. This,
patch fixes this logic to match the resolver.

Epic: none

Release note: None

fqazi committed:
sql/schemachanger: add support for the CREATE SEQUENCE

Previously, the CREATE SEQUENCE command only worked
in the legacy schema changer. This patch, adds support
for creating sequences in declarative schema changer.

Epic: None
Release note: None

fqazi committed:
sql/schemachanger: disable two version invariant for added descriptors

Previously, the two version invariant was enforced
for all non-dropped descriptors. When creating new
descriptors if they are going to be decomposed into
individual elements, then similar logic needs to
skip added descriptors. To address this, this patch
disables these rules for added descriptors.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104350)
<!-- Reviewable:end -->

postamar commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_column_not_null.go:
```diff
@@ -28,7 +28,10 @@ func init() {
+				emit(func(this *scpb.ColumnNotNull, md *opGenContext) *scop.ValidateColumnNotNull {
+					if checkIfDescriptorIsAdded(this.TableID, md) {
+						return nil
+					}
```
I don't think that this works. Assuming that CREATE TABLE was supported, consider
```
BEGIN;
CREATE TABLE t (x INT PRIMARY KEY);
INSERT INTO t VALUES (1);
ALTER TABLE t ADD COLUMN y INT NOT NULL DEFAULT 42;
COMMIT;
```
We still need to backfill and validate in that case, despite the relation descriptor being added.

What we really want is backfills and validation being conditional on the descriptor being added _and_ the relation being non-empty. We do have TableData and IndexData elements to model the presence of data. Right now these are used for DROP and we just assume that the data always exists in any case, but it needn't be so.

Anyway that's out of scope but `checkIfDescriptorIsAdded` should be expanded to also check for the presence of a ToPublic `TableData` or `IndexData` target, and those targets need to be set by the builder as well.

fqazi replied:
Good point. I added a check for this but, for the longer term we need to figure out how to decompose emptiness in terms of round trips and everything (feels like something in collections)

postamar commented on pkg/sql/schemachanger/scpb/elements.proto:
```diff
@@ -304,10 +308,23 @@ message SchemaChild {
+  int64 cache_size = 6;
+  bool virtual = 7;
+  string as_integer_type = 8;
+}
```
This won't work when we have to implement `ALTER SEQUENCE`. How would you set targets for
```
BEGIN;
ALTER SEQUENCE seq MINVALUE 10;
ALTER SEQUENCE seq MAXVALUE 100;
COMMIT;
```

Instead, consider adding a `SequenceOption` element:
```
message SequenceOptions {
  uint32 sequence_id = 1 [(gogoproto.customname) = "SequenceID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
  string key = 2;
  string value = 3;
}
```
and bind the `DescID` attr to the `SequenceID` field and the `Name` attr to `Key` field.

fqazi replied:
Good point, I'll refresh the code shortly with that.

postamar reviewed (commented):
I took an early look at the last 3 commits and left some preliminary comments. Nicely done. Can you add support for ADD COLUMN SERIAL and friends, if that's straightforward? I can't remember too well, but it might be. Don't bother if it's out of scope, but it'd be a nice test case.

fqazi committed:
sql/schemachanger: add support for OWNED BY on CREATE SEQUENCE

Previously, our CREATE SEQUENCE support for the declarative
schema changer fell back if an OWNED BY clause is specified.
To address this patch adds support for owned by clause.

Epic: none

Release note: None

fqazi commented:
> I took an early look at the last 3 commits and left some preliminary comments. Nicely done. Can you add support for ADD COLUMN SERIAL and friends, if that's straightforward? I can't remember too well, but it might be. Don't bother if it's out of scope, but it'd be a nice test case.

We have a tracking issue for the remaining ADD COLUMN work, but I'll definitely circle back as a follow on PR

fqazi committed:
sql/schemachanger: emit data elements for schema changes

To support the ability of skipping operations when data
exists, we will add support for including data targets
within plans even if they have no transitions. These
elements then can be used by rules to decide if backfill
or the two version invariant are required. This patch,
will add logic to do this detection.

Epic: none

Release note: None

fqazi committed:
sql/schemachanger: skip validation and backfill for added tables

Previously, we would always generate operations to backfill
and validate on newly created tables (with data elements).
This will prevent us from being able to generate single stage
plans for creating relations. This patch removes those checks
for added tables, and allows data elements to be emitted as
targets.

Epic: none
Release note: None

fqazi commented:
@postamar This is ready for another look, its rebased and now uses TableData elements for decision making so more future proof.

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/build.go:
```diff
@@ -104,10 +104,16 @@ func Build(
+	}
 	for _, e := range bs.output {
-		if e.metadata.Size() == 0 {
+		if e.metadata.Size() == 0 && !shouldElementBeRetainedWithoutMetadata(e.element) {
```
Please help me understand the overall purpose of including those `TableData` element and `IndexData` element in the TargetState even if they are not added/dropped in the builder. In other words, how exactly do they help "hinting" if a backfill is needed

fqazi replied:
They allow us to know if a table is empty, which allows us to skip backfill steps completely. The other alternative is to use the DESCRIPTOR_ADDED state, but that breaks if we enter a more transactional world. Since you should be able to insert on those

fqazi replied:
Updated the comments here:
			// Exceptions are TableData/IndexData elements which allow our planning
			// execution to skip certain transitions and fences like the two version
			// invariant. We will only keep these for descriptors going through
			// a transition.

fqazi committed:
sql/schemachanger: add migration to inject table data elements

Previously, if an existing job was missing the TableData
element, the new job would not properly handle the current
set of rules in the mixed version state. This patch adds,
a migration to inject those elements.

Epic: none

Release note: None

fqazi commented on pkg/sql/schemachanger/scbuild/build.go:
```diff
@@ -121,12 +127,32 @@ func Build(
+		}
+
+	}
+	seenDescriptors := screl.AllTargetDescIDs(ts)
+	// Elements without metadata can only be retained, if some other descriptor
+	// is referencing them. We may have multiple index data and table data's,
+	// but not all of them are relevant if the primary descriptor ID never shows.
+	for _, ex := range extraTargets {
+		if seenDescriptors.Contains(screl.GetDescID(ex.t.Element())) {
+			ts.Targets = append(ts.Targets, ex.t)
+			initial = append(initial, ex.e.initial)
+			current = append(current, ex.e.current)
```
// We are going to retain certain elements for metadata purposes, like
	// TableData/IndexData elements, which will allow the declarative schema
	// changer to know if a given table is empty. Once these elements exist
	// the two-version invariant is applied when making mutations to elements.
	// Only emit data elements for descriptors if they are references with
	// some transition.

postamar commented on pkg/sql/catalog/rewrite/rewrite.go:
```diff
@@ -651,6 +651,16 @@ func rewriteSchemaChangerState(
+			data.TableID = rewrite.ID
+			data.DatabaseID = rewrite.ParentID
+			continue
+		}
```
This rewrite makes sense but should this be in the switch case below? Not a rethorical question. Or is this case somehow different? If so I can't see it and this deserves commentary perhaps.

fqazi replied:
It's because the parentID is not contained within the map that we are using, so it ends up special cased. Let me add the comment:
		// Since the parent database ID is never written in the descriptorRewrites
		// map we need to special case certain elements that need their ParentID
		// re-written.

postamar commented on pkg/sql/schemachanger/scpb/state.go:
```diff
@@ -141,6 +146,11 @@ func (e *ElementProto) Element() Element {
 }
 
+// TargetIsLinkedToSchemaChange return if a Target is linked to a schema change.
+func (t *Target) TargetIsLinkedToSchemaChange() bool {
```
I like this.

postamar commented on pkg/sql/schemachanger/scpb/state.go:
```diff
@@ -94,6 +94,11 @@ func (s *CurrentState) Rollback() {
 		t := &s.Targets[i]
+		// If the metadata is not populated this element
+		// only usd for tracking.
+		if t.Metadata.Size() == 0 {
```
Call `TargetIsLinkedToSchemaChange` instead?

postamar commented on pkg/sql/schemachanger/scrun/scrun.go:
```diff
@@ -273,7 +273,8 @@ func logSchemaChangeEvents(
 	for _, t := range state.TargetState.Targets {
-		if t.Metadata.SourceElementID > 1 {
+		if t.Metadata.SourceElementID > 1 ||
+			t.Metadata.Size() == 0 { // Ignore empty metadata
```
Call `TargetIsLinkedToSchemaChange` instead?

postamar commented on pkg/sql/schemachanger/scbuild/build.go:
```diff
@@ -104,10 +104,18 @@ func Build(
+	}
 	for _, e := range bs.output {
-		if e.metadata.Size() == 0 {
+		if e.metadata.Size() == 0 && !shouldElementBeRetainedWithoutMetadata(e.element) {
```
Call `TargetIsLinkedToSchemaChange` instead?

postamar commented on pkg/sql/schemachanger/scbuild/build.go:
```diff
@@ -441,3 +471,16 @@ func (b buildCtx) WithNewSourceElementID() scbuildstmt.BuildCtx {
+		return true
+	}
+	return false
+}
```
I wonder if perhaps we'd be better served by adding a boolean override flag in the metadata protobuf instead, that would be set by `scdecomp` for TableData/IndexData elements. I'm not sure if it's actually better, perhaps not. Your approach is fine in the meantime, perhaps I'll explore this after this merges.

postamar commented on pkg/sql/catalog/dbdesc/database_desc_builder.go:
```diff
@@ -162,7 +162,7 @@ func (ddb *databaseDescriptorBuilder) RunRestoreChanges(
 ) error {
 	// Upgrade the declarative schema changer state.
-	if scpb.MigrateDescriptorState(version, ddb.maybeModified.DeclarativeSchemaChangerState) {
+	if scpb.MigrateDescriptorState(version, ddb.maybeModified.ID, ddb.maybeModified.DeclarativeSchemaChangerState) {
```
Shouldn't the second arg be 0?

postamar commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_column_not_null.go:
```diff
@@ -28,7 +28,10 @@ func init() {
+				emit(func(this *scpb.ColumnNotNull, md *opGenContext) *scop.ValidateColumnNotNull {
+					if checkIfDescriptorIsWithoutData(this.TableID, md) {
+						return nil
+					}
```
What about other constraints? Shouldn't they also have this check?

postamar commented on pkg/sql/schemachanger/scplan/internal/scstage/build.go:
```diff
@@ -320,12 +320,29 @@ func (bc buildContext) makeStageBuilderForType(bs buildState, opType scop.Type)
+				} else {
+					break
+				}
+			}
```
This change makes sense but I'm a bit concerned about the added computational complexity that this loop adds. Can we pre-compute the `anyRemainingOpsCanFail` property for each node instead? Define `anyRemainingOpsCanFail := make(map[*screl.Node]bool), then for each target, start with the sink node and do a depth-first traversal going backwards along each op edge, if you've verified `oe.CanFail()` at least once in the path to the sink node then set `anyRemainingOpsCanFail[oe.From()] = true` and vice-versa.

postamar commented on pkg/sql/logictest/testdata/logic_test/crdb_internal:
```diff
@@ -861,7 +863,7 @@ start_pretty    end_pretty
 /Table/112/2/1  /Table/112/2/2
 /Table/112/2/2  /Table/112/2/3
-/Table/112/2/3  /Max
+/Table/112/2/3  /Table/113/1
```
Remind me, what triggered this change?

postamar commented on pkg/sql/schemachanger/scbuild/testdata/unimplemented_create:
```diff
@@ -14,10 +14,6 @@ unimplemented
-unimplemented
-CREATE SEQUENCE seq;
-----
-
```
Hooray :)

postamar commented on pkg/sql/schemachanger/scexec/dependencies.go:
```diff
@@ -96,6 +97,13 @@ type Catalog interface {
+type RelationHelpers interface {
+	// InitializeSequence initializes the initial value for a sequence.
+	InitializeSequence(id descpb.ID, startVal int64)
+}
```
This is fine, though FYI I wouldn't have minded simply adding the `InitializeSequence` method to `scexec.Catalog`, because I'm not sure how much this interface will grow in the future.

postamar commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_sequence.go:
```diff
@@ -21,12 +21,27 @@ func init() {
+					}
+					return &scop.InitSequence{
+						SequenceID:  this.SequenceID,
+						RestartWith: restartWtih,
```
Instead of having a pointer to an int, can we have an extra `UseRestartWith` boolean field instead? I'm not sure how well an *int64 serializes in YAML (which is how we render EXPLAIN diagrams) and this isn't tested. For consistency, you might want to do the same in the element model.

postamar commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_sequence_option.go:
```diff
@@ -0,0 +1,41 @@
+// Copyright 2021 The Cockroach Authors.
```
We're in 2023.

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -229,11 +231,36 @@ func (w *walkCtx) walkRelation(tbl catalog.TableDescriptor) {
+				if value == nil {
+					return
+				}
+				keyStr := fmt.Sprintf("%v", value)
```
s/keyStr/valueStr/

postamar commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -229,11 +231,36 @@ func (w *walkCtx) walkRelation(tbl catalog.TableDescriptor) {
+				keyStr := fmt.Sprintf("%v", value)
+				if len(keyStr) == 0 {
+					return
+				}
```
You need to also return when the value is 0 (or whatever the default is) for integer values. Perhaps pass an extra `defaultValue` argument to `addSequenceOption` and compare the value to that?

postamar commented on pkg/sql/schemachanger/testdata/explain/create_sequence:
```diff
@@ -0,0 +1,130 @@
+ │         │    ├── ABSENT → PUBLIC SequenceOption:{DescID: 104 (sq1+), Name: "MAXVALUE"}
+ │         │    ├── ABSENT → PUBLIC SequenceOption:{DescID: 104 (sq1+), Name: "START"}
+ │         │    ├── ABSENT → PUBLIC SequenceOption:{DescID: 104 (sq1+), Name: "VIRTUAL"}
+ │         │    ├── ABSENT → PUBLIC SequenceOption:{DescID: 104 (sq1+), Name: "CACHE"}
```
This is out of scope for this PR but come to think of it, we'll also have to bind the value to a new `Value` attr in screl at some point, to handle `ALTER SEQUENCE` DDLs.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_sequence.go:
```diff
@@ -100,6 +100,14 @@ func CreateSequence(b BuildCtx, n *tree.CreateSequence) {
+		DescriptorID: sequenceID,
+		Name:         string(n.Name.ObjectName),
+	}
+	b.Add(sequenceNamespace)
```
This seems like it should be in the previous commit.

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/references.go:
```diff
@@ -71,6 +71,33 @@ func (i *immediateVisitor) RemoveOwnerBackReferenceInSequence(
+	opts.SequenceOwner.OwnerColumnID = op.ColumnID
+	opts.SequenceOwner.OwnerTableID = op.TableID
+	return nil
+}
```
nit: missing newline

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/references.go:
```diff
@@ -71,6 +71,33 @@ func (i *immediateVisitor) RemoveOwnerBackReferenceInSequence(
+		return err
+	}
+	col, err := catalog.MustFindColumnByID(tbl, op.ColumnID)
+	if err != nil || col == nil {
```
nit: `col == nil` check is superfluous

postamar reviewed (changes requested):
Nicely done! I only have minor complaints.

postamar commented:
It doesn't seem that all my comments have been addressed. If you'd rather not, can you please respond with why? Thanks

fqazi reviewed (commented):
@postamar Sorry, that wasn't intentional. There are a bunch which were on Reviewable that I didn't notice somehow, let me get those.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104350#-:-NYsDNn41h_Y3DRSzIvq:bxs36vv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek, @postamar, and @Xiang-Gu)



<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/catalog/dbdesc/database_desc_builder.go` line 165 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231007870:-NYsEyvc6yoP5jBAvkgH:b-oz8o77) ([raw file](https://github.com/cockroachdb/cockroach/blob/3492000f09bb57d1954b865685ba5d6dc2b3f9eb/pkg/sql/catalog/dbdesc/database_desc_builder.go#L165)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't the second arg be 0?
</blockquote></details>

Good catch
___
*[`pkg/sql/schemachanger/scbuild/build.go` line 112 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1230989306:-NYsF9I36OVRxMIrKju_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0649fd63d953d6a611cfeecbb4f5ca5309225df3/pkg/sql/schemachanger/scbuild/build.go#L112)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Call `TargetIsLinkedToSchemaChange` instead?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/build.go` line 486 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231000981:-NYu_mK69NiCI7SCUYia:bdvr6un) ([raw file](https://github.com/cockroachdb/cockroach/blob/32a29271d13b8aa49e46e640027ce8c4a412c0ba/pkg/sql/schemachanger/scbuild/build.go#L486)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I wonder if perhaps we'd be better served by adding a boolean override flag in the metadata protobuf instead, that would be set by `scdecomp` for TableData/IndexData elements. I'm not sure if it's actually better, perhaps not. Your approach is fine in the meantime, perhaps I'll explore this after this merges.
</blockquote></details>

I think that is pretty reasonable and would generalize this a lot better.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 245 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231051747:-NYuaJXL4s-Tsq0ZVlFf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/schemachanger/scdecomp/decomp.go#L245)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

s/keyStr/valueStr/
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 248 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231053648:-NYua8PxA3cO_8VVlhFo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/schemachanger/scdecomp/decomp.go#L248)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You need to also return when the value is 0 (or whatever the default is) for integer values. Perhaps pass an extra `defaultValue` argument to `addSequenceOption` and compare the value to that?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scexec/dependencies.go` line 105 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231039718:-NYub8CkEXTorGUbFOoB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/schemachanger/scexec/dependencies.go#L105)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is fine, though FYI I wouldn't have minded simply adding the `InitializeSequence` method to `scexec.Catalog`, because I'm not sure how much this interface will grow in the future.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scpb/state.go` line 99 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1230987189:-NYsFPGlCbxerXxvk0Kp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0649fd63d953d6a611cfeecbb4f5ca5309225df3/pkg/sql/schemachanger/scpb/state.go#L99)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Call `TargetIsLinkedToSchemaChange` instead?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scpb/state.go` line 150 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1230986823:-NYsFPpl6hI-wzCYg7FE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0649fd63d953d6a611cfeecbb4f5ca5309225df3/pkg/sql/schemachanger/scpb/state.go#L150)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I like this.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_column_not_null.go` line 34 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231010676:-NYsGfV38U6h0yLe3DIZ:bwnn4s5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d5224f9d8f9a3f9e9561b40c970ed5269630321/pkg/sql/schemachanger/scplan/internal/opgen/opgen_column_not_null.go#L34)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

What about other constraints? Shouldn't they also have this check?
</blockquote></details>

Yes, they, need the same check let me update them as well.

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_sequence.go` line 42 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231049366:-NYsMxcg47Qvk6aDSnkE:b-fuw5po) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/schemachanger/scplan/internal/opgen/opgen_sequence.go#L42)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Instead of having a pointer to an int, can we have an extra `UseRestartWith` boolean field instead? I'm not sure how well an *int64 serializes in YAML (which is how we render EXPLAIN diagrams) and this isn't tested. For consistency, you might want to do the same in the element model.
</blockquote></details>

Good point, that's much cleaner. The element model can have a similar change too.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_sequence_option.go` line 1 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231050061:-NYsFUwn9JVaGEQe5xGN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/schemachanger/scplan/internal/opgen/opgen_sequence_option.go#L1)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

We're in 2023.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/scstage/build.go` line 345 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231027333:-NYucDdQA4sz_vZ727Lt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d5224f9d8f9a3f9e9561b40c970ed5269630321/pkg/sql/schemachanger/scplan/internal/scstage/build.go#L345)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This change makes sense but I'm a bit concerned about the added computational complexity that this loop adds. Can we pre-compute the `anyRemainingOpsCanFail` property for each node instead? Define `anyRemainingOpsCanFail := make(map[*screl.Node]bool)`, then for each target, start with the sink node and do a depth-first traversal going backwards along each op edge, if you've verified `oe.CanFail()` at least once in the path to the sink node then set `anyRemainingOpsCanFail[oe.From()] = true` and vice-versa.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scrun/scrun.go` line 277 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1230987262:-NYsF_3w07elBss5M5Ev:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0649fd63d953d6a611cfeecbb4f5ca5309225df3/pkg/sql/schemachanger/scrun/scrun.go#L277)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Call `TargetIsLinkedToSchemaChange` instead?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal` line 866 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231029535:-NYusTeLCq0wuBcbB_Vg:b-baicqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/logictest/testdata/logic_test/crdb_internal#L866)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Remind me, what triggered this change?
</blockquote></details>

This is triggered by MaybeSplit.* from the primary index. I guess the question is if we want this temporary admin split or not since for a sequence it's only a single row.
___
*[`pkg/sql/schemachanger/scbuild/testdata/unimplemented_create` line 0 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231030490:-NYsFbWl24ssXOznpd1G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/schemachanger/scbuild/testdata/unimplemented_create#L0)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

Hooray :)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi commented:
@postamar ready for another look

postamar commented on pkg/sql/schemachanger/testdata/explain_verbose/create_sequence:
```diff
@@ -0,0 +1,778 @@
+            │     SequenceID: 104
+            │
+            └── • MarkDescriptorAsPublic
+                  DescriptorID: 104
```
This file shouldn't exist.

postamar commented on pkg/sql/schemachanger/testdata/end_to_end/drop_column_create_index_separate_statements/create_sequence:
```diff
@@ -0,0 +1,130 @@
+                ├── SetIndexName {"IndexID":1,"Name":"primary","TableID":104}
+                ├── MakeValidatedPrimaryIndexPublic {"IndexID":1,"TableID":104}
+                ├── InitSequence {"SequenceID":104}
+                └── MarkDescriptorAsPublic {"DescriptorID":104}
```
This file shouldn't exist.

postamar commented on pkg/sql/schemachanger/testdata/end_to_end/create_schema_drop_schema_separate_statements/create_sequence:
```diff
@@ -0,0 +1,7 @@
+EXPLAIN (ddl, shape) CREATE SEQUENCE sq1 MINVALUE 1 MAXVALUE 9223372036854775807 INCREMENT 1 START 32;
+----
+Schema change plan for CREATE SEQUENCE ‹defaultdb›.‹public›.‹sq1› MINVALUE 1 MAXVALUE 9223372036854775807 INCREMENT 1 START 32;
+ └── execute 1 system table mutations transaction
```
This file shouldn't exist.

postamar reviewed (changes requested):
You're missing at least one end-to-end test case involving `CREATE SEQUENCE`.

postamar reviewed (approved):
At this point if the CI's happy I'm happy. I've looked at enough iterations of this PR that I'm starting to lose track of things. I don't want to block this, though, and I'm feeling pretty good about this change.. Feel free to have this looked at by a fresh pair of eyes, depending on how confident you are about this change. It's up to you.

fqazi commented:
> At this point if the CI's happy I'm happy. I've looked at enough iterations of this PR that I'm starting to lose track of things. I don't want to block this, though, and I'm feeling pretty good about this change.. Feel free to have this looked at by a fresh pair of eyes, depending on how confident you are about this change. It's up to you.

Thanks for the patience in reviewing this, and I appreciate the feedback it made the entire thing much better. I'm gonna make sure CI is clear then merge it with the remaining fixes.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104350)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek, @postamar, and @Xiang-Gu)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1246611158:-NZ6LYU57uWeR1QDrvjU:b-896fix) (no related file):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

You don't actually need an expected output in the .definition file. Please remove.
</blockquote></details>

Done.

___
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_sequence.go` line 110 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231059152:-NZ2BdW-DKSrFVRTKQFH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32a29271d13b8aa49e46e640027ce8c4a412c0ba/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_sequence.go#L110)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This seems like it should be in the previous commit.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_sequence.go` line 135 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1245571839:-NZ3zntf77aGSUvcXl5i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44153a6bc52a1201c819eed3933c75bbfb44ba94/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_sequence.go#L135)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't you be reusing the same logic as in scdecomp here for the sequence options?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 245 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1231051747:-NZ2BuDQBXcE9kFJq6yN:byhmem0) ([raw file](https://github.com/cockroachdb/cockroach/blob/410a1e43526f34d1d68b2285bb5c9d5b6324173e/pkg/sql/schemachanger/scdecomp/decomp.go#L245)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This hasn't been fixed.
</blockquote></details>

Ugh, I fixed the comment but forgot to fix the variable name.
___
*[`pkg/sql/schemachanger/scplan/internal/scstage/build.go` line 100 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1245576288:-NZ6JjGF7F854zTjDb_g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44153a6bc52a1201c819eed3933c75bbfb44ba94/pkg/sql/schemachanger/scplan/internal/scstage/build.go#L100)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Please move this logic into a top-level function which you'll call in the struct initialization above: `bc := buildContext{ ... }`.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/create_schema_drop_schema_separate_statements/create_sequence` line 7 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1245579809:-NZ3zlIN-FDb6ZCs1Jlr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44153a6bc52a1201c819eed3933c75bbfb44ba94/pkg/sql/schemachanger/testdata/end_to_end/create_schema_drop_schema_separate_statements/create_sequence#L7)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This file shouldn't exist.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/drop_column_create_index_separate_statements/create_sequence` line 130 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1245579579:-NZ3zmA9D0y76LAjPAtt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44153a6bc52a1201c819eed3933c75bbfb44ba94/pkg/sql/schemachanger/testdata/end_to_end/drop_column_create_index_separate_statements/create_sequence#L130)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This file shouldn't exist.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/testdata/explain_verbose/create_sequence` line 778 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/104350#gh-1245579144:-NZ3zmpi50rTi7jmSKYU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44153a6bc52a1201c819eed3933c75bbfb44ba94/pkg/sql/schemachanger/testdata/explain_verbose/create_sequence#L778)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This file shouldn't exist.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+
